src/anyp/Uri.cc

```diff
@@ -81,6 +103,50 @@ AnyP::Uri::Encode(const SBuf &buf, const CharacterSet &ignore)
     return output;
 }
 
+SBuf
+AnyP::Uri::Rfc3986Encode(const SBuf &buf)
+{
+    return Encode(buf, CharacterSet::RFC3986_UNRESERVED);
+}
+
+SBuf
+AnyP::Uri::Decode(const SBuf &buf)
+{
+    if (buf.isEmpty())
+        return buf;
+
+    SBuf output;
+    output.reserveSpace(buf.length()); // worst case: buf is not encoded
+
+    Parser::Tokenizer tok(buf);
+    SBuf token;
+    static const auto unencodedOctets = CharacterSet("unencoded", "%").complement();
+
+    tok.prefix(token, unencodedOctets);
+    output.append(token);
+
+    while (!tok.atEnd())
+    {
+        const auto foundPercent = tok.skip('%');
+        Assure2(foundPercent, "Error decoding Uri: '%' expected but not found");
+
+        if (!tok.prefix(token, HexDigits, 2) || token.length() != 2)
+            throw TextException("incomplete or invalid %-encoded triplet", Here());
+
+        const auto hex1 = HextDigitToInt(token[0]);
+        const auto hex2 = HextDigitToInt(token[1]);
+        if (hex1 == -1 || hex2 == -1)
+            throw TextException("invalid %-encoded triplet", Here());
+
+        const char decoded = hex1 << 4 | hex2;
+        output.append(decoded);
```

Comment:
`Tokenizer` has built-in hex logic. We can use that instead of inventing a new `HexDigitToIntMap` algorithm.
In order to require exactly-2 characters we do have to parse them individually, placing a limit of 1 handles that fine.

Refactor step 4:
```suggestion
            int64_t hex1 = 0, hex2 = 0;
            if (tok.int64(hex1, 16, false, 1) && tok.int64(hex2, 16, false, 1)) {
                output.append(char((hex1<<4)|hex2));
            } else
                throw TextException("incomplete or invalid %-encoded triplet", Here());
```

Reply:
Adopted. Thanks!